File: client/src/app/crm/features/emails/services/crm-email-link.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../../../environments/environment';
import { CrmLinkEntityType, CrmEmailLink, CrmEmailLinkRequest } from '../models/email.model';

export interface CrmEmailLinkResponse {
  id: string;
  connectionId: string;
  externalMessageId: string;
  conversationId?: string;
  subject: string;
  fromEmail: string;
  fromName?: string;
  receivedAtUtc: string;
  provider: string;
  relatedEntityType: string;
  relatedEntityId: string;
  linkedByUserId: string;
  note?: string;
  createdAtUtc: string;
}

@Injectable({ providedIn: 'root' })
export class CrmEmailLinkService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = environment.apiUrl;

  linkEmail(request: CrmEmailLinkRequest): Observable<CrmEmailLinkResponse> {
    return this.http.post<CrmEmailLinkResponse>(`${this.baseUrl}/api/mailbox/links`, request);
  }

  unlinkEmail(linkId: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/api/mailbox/links/${linkId}`);
  }

  getLinksForEntity(entityType: CrmLinkEntityType, entityId: string): Observable<CrmEmailLinkResponse[]> {
    return this.http.get<CrmEmailLinkResponse[]>(
      `${this.baseUrl}/api/mailbox/links/entity/${entityType}/${entityId}`
    );
  }

  getMyLinks(): Observable<CrmEmailLinkResponse[]> {
    return this.http.get<CrmEmailLinkResponse[]>(`${this.baseUrl}/api/mailbox/links/mine`);
  }
}

File: server/src/CRM.Enterprise.Infrastructure/Emails/CrmEmailLinkService.cs
using CRM.Enterprise.Application.Emails;
using CRM.Enterprise.Domain.Entities;
using CRM.Enterprise.Domain.Enums;
using CRM.Enterprise.Infrastructure.Persistence;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

namespace CRM.Enterprise.Infrastructure.Emails;

public sealed class CrmEmailLinkService : ICrmEmailLinkService
{
    private readonly CrmDbContext _dbContext;
    private readonly ILogger<CrmEmailLinkService> _logger;

    public CrmEmailLinkService(CrmDbContext dbContext, ILogger<CrmEmailLinkService> logger)
    {
        _dbContext = dbContext;
        _logger = logger;
    }

    public async Task<CrmEmailLinkDto> LinkEmailAsync(CreateCrmEmailLinkRequest request, CancellationToken ct = default)
    {
        // Check if already linked
        var existing = await _dbContext.CrmEmailLinks
            .FirstOrDefaultAsync(l =>
                l.ConnectionId == request.ConnectionId &&
                l.ExternalMessageId == request.ExternalMessageId &&
                l.RelatedEntityType == request.RelatedEntityType &&
                l.RelatedEntityId == request.RelatedEntityId &&
                !l.IsDeleted, ct);

        if (existing is not null)
            return MapToDto(existing);

        var connection = await _dbContext.UserEmailConnections
            .FirstOrDefaultAsync(c => c.Id == request.ConnectionId && !c.IsDeleted, ct)
            ?? throw new InvalidOperationException($"Email connection {request.ConnectionId} not found.");

        var link = new CrmEmailLink
        {
            ConnectionId = request.ConnectionId,
            ExternalMessageId = request.ExternalMessageId,
            ConversationId = request.ConversationId,
            Subject = request.Subject,
            FromEmail = request.FromEmail,
            FromName = request.FromName,
            ReceivedAtUtc = request.ReceivedAtUtc,
            Provider = connection.Provider,
            RelatedEntityType = request.RelatedEntityType,
            RelatedEntityId = request.RelatedEntityId,
            LinkedByUserId = request.LinkedByUserId,
            Note = request.Note
        };

        _dbContext.CrmEmailLinks.Add(link);
        await _dbContext.SaveChangesAsync(ct);

        _logger.LogInformation(
            "Email {ExternalMessageId} linked to {EntityType} {EntityId} by user {UserId}",
            request.ExternalMessageId, request.RelatedEntityType, request.RelatedEntityId, request.LinkedByUserId);

        return MapToDto(link);
    }

    public async Task<bool> UnlinkEmailAsync(Guid linkId, CancellationToken ct = default)
    {
        var link = await _dbContext.CrmEmailLinks
            .FirstOrDefaultAsync(l => l.Id == linkId && !l.IsDeleted, ct);

        if (link is null) return false;

        link.IsDeleted = true;
        link.DeletedAtUtc = DateTime.UtcNow;
        await _dbContext.SaveChangesAsync(ct);

        _logger.LogInformation("Email link {LinkId} removed", linkId);
        return true;
    }

    public async Task<IReadOnlyList<CrmEmailLinkDto>> GetLinksForEntityAsync(
        ActivityRelationType entityType, Guid entityId, CancellationToken ct = default)
    {
        var links = await _dbContext.CrmEmailLinks
            .Where(l => l.RelatedEntityType == entityType && l.RelatedEntityId == entityId && !l.IsDeleted)
            .OrderByDescending(l => l.ReceivedAtUtc)
            .ToListAsync(ct);

        return links.Select(MapToDto).ToList();
    }

    public async Task<IReadOnlyList<CrmEmailLinkDto>> GetLinksForUserAsync(Guid userId, CancellationToken ct = default)
    {
        var links = await _dbContext.CrmEmailLinks
            .Where(l => l.LinkedByUserId == userId && !l.IsDeleted)
            .OrderByDescending(l => l.ReceivedAtUtc)
            .ToListAsync(ct);

        return links.Select(MapToDto).ToList();
    }

    private static CrmEmailLinkDto MapToDto(CrmEmailLink link) => new(
        link.Id,
        link.ConnectionId,
        link.ExternalMessageId,
        link.ConversationId,
        link.Subject,
        link.FromEmail,
        link.FromName,
        link.ReceivedAtUtc,
        link.Provider.ToString(),
        link.RelatedEntityType.ToString(),
        link.RelatedEntityId,
        link.LinkedByUserId,
        link.Note,
        link.CreatedAtUtc
    );
}
